consul/client.go

```diff
@@ -0,0 +1,509 @@
+		retry, err = c.reconnectIfConnectionError(conn, err)
+		return retry
+	})
+
+	return list, meta, err
+}
+
+func (c *consulClientImpl) List(prefix string, q *api.QueryOptions) (api.KVPairs, *api.QueryMeta, error) {
+	var pairs api.KVPairs
+	var meta *api.QueryMeta
+	var err error
+	retry := false
+
+	c.runWithRetry(func() bool {
+		conn := c.conn
+		pairs, meta, err = conn.client.KV().List(prefix, nil)
+		retry, err = c.reconnectIfConnectionError(conn, err)
+		return retry
+	})
+
+	return pairs, meta, err
+}
+
+func (c *consulClientImpl) Acquire(p *api.KVPair, q *api.WriteOptions) (*api.WriteMeta, error) {
+	var err error
+	var meta *api.WriteMeta
+	var ok bool
+	retry := false
+	c.runWithRetry(func() bool {
+		conn := c.conn
+		ok, meta, err = conn.client.KV().Acquire(p, nil)
+		retry, err = c.reconnectIfConnectionError(conn, err)
+		return retry
+	})
+
+	// *** this error is created in loop above
+	if err != nil {
+		return nil, err
+	}
+
+	if !ok {
+		return nil, fmt.Errorf("Acquire failed")
+	}
+
+	return meta, err
+}
+
+func (c *consulClientImpl) Create(se *api.SessionEntry, q *api.WriteOptions) (string, *api.WriteMeta, error) {
+	var session string
+	var meta *api.WriteMeta
+	var err error
+	retry := false
+
+	c.runWithRetry(func() bool {
+		conn := c.conn
+		session, meta, err = conn.client.Session().Create(se, q)
+		retry, err = c.reconnectIfConnectionError(conn, err)
+		return retry
+	})
+
+	return session, meta, err
+}
+
+func (c *consulClientImpl) Destroy(id string, q *api.WriteOptions) (*api.WriteMeta, error) {
+	return c.writeRetryFunc(func(conn *consulConnection) (*api.WriteMeta, error) {
+		return conn.client.Session().Destroy(id, q)
+	})
+}
+
+func (c *consulClientImpl) Renew(id string, q *api.WriteOptions) (*api.SessionEntry, *api.WriteMeta, error) {
+	var entry *api.SessionEntry
+	var meta *api.WriteMeta
+	var err error
+	retry := false
+
+	c.runWithRetry(func() bool {
+		conn := c.conn
+		entry, meta, err = conn.client.Session().Renew(id, q)
+		retry, err = c.reconnectIfConnectionError(conn, err)
+		return retry
+	})
+
+	return entry, meta, err
+}
+
+func (c *consulClientImpl) RenewPeriodic(
+	initialTTL string,
+	id string,
+	q *api.WriteOptions,
+	doneCh chan struct{},
+) error {
+	var err error
+	retry := false
+
+	c.runWithRetry(func() bool {
+		conn := c.conn
+		err = conn.client.Session().RenewPeriodic(initialTTL, id, nil, doneCh)
+		retry, err = c.reconnectIfConnectionError(conn, err)
+		return retry
+	})
+
+	return err
+}
+
+func (c *consulClientImpl) CreateMeta(
+	id string,
+	p *api.KVPair,
+	q *api.WriteOptions,
+) (*api.WriteMeta, bool, error) {
+	var ok bool
+	var meta *api.WriteMeta
+	var err error
+	connError := false
+
+	for i := 0; i < c.maxRetries; i++ {
+		conn := c.conn
+		ok, meta, err = conn.client.KV().Acquire(p, q)
+		if ok && err == nil {
+			return nil, ok, err
+		}
+		if _, err := conn.client.Session().Destroy(p.Session, nil); err != nil {
```

Comment:
why are we doing the destroy and delete operations with the old connection before reconnecting?